server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java

```diff
@@ -8,36 +8,51 @@
 
 package org.elasticsearch.search.aggregations.bucket.composite;
 
+import org.apache.lucene.index.DocValues;
 import org.apache.lucene.index.IndexReader;
 import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PostingsEnum;
 import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
 import org.apache.lucene.search.MatchAllDocsQuery;
 import org.apache.lucene.search.Query;
 import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.PriorityQueue;
 import org.elasticsearch.common.util.BigArrays;
 import org.elasticsearch.common.util.LongArray;
 import org.elasticsearch.core.CheckedFunction;
 import org.elasticsearch.core.Releasables;
 import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.index.mapper.StringFieldType;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
 import org.elasticsearch.search.DocValueFormat;
 import org.elasticsearch.search.aggregations.LeafBucketCollector;
 
 import java.io.IOException;
+import java.util.ArrayDeque;
 
 import static org.apache.lucene.index.SortedSetDocValues.NO_MORE_ORDS;
 
 /**
  * A {@link SingleDimensionValuesSource} for global ordinals.
  */
 class GlobalOrdinalValuesSource extends SingleDimensionValuesSource<BytesRef> {
+
+    private static final Logger logger = LogManager.getLogger(GlobalOrdinalValuesSource.class);
+
     public static final long MISSING_VALUE_FLAG = -1L;
     private final CheckedFunction<LeafReaderContext, SortedSetDocValues, IOException> docValuesFunc;
     private LongArray values;
     private SortedSetDocValues lookup;
     private long currentValue;
     private Long afterValueGlobalOrd;
+    private Long highestCompetitiveValueGlobalOrd = MISSING_VALUE_FLAG;
     private boolean isTopValueInsertionPoint;
+    private volatile CompetitiveIterator currentCompetitiveIterator;
```

Comment:
I believe this could cause issues with how TimeSeriesIndexSearcher is collecting mutiple segments in parallel. Or maybe `composite` cannot be used with `composite` anyway? Maybe @martijnvg has thoughts on this.

Reply:
The `composite` aggregation sometimes doesn't work with the `TimeSeriesIndexSearcher`.

In the cases that the `getLeafCollector(...)` returns `LeafBucketCollector.NO_OP_COLLECTOR` then search execution fails with a `collection_terminated_exception` error. In other cases composite does seem to work with `TimeSeriesIndexSearcher`. We should look into how to fix this. The problem is related why sample agg or the filter by filter optimisation doesn't work. 

However if we make this change then there is another reason why `composite` aggregation doesn't work and that is because `currentCompetitiveIterator` will likely refer to the wrong `CompetitiveIterator` instance. The `TimeSeriesIndexSearcher` creates `LeafWalker` (with `LeafBucketCollector` instance) instances for all segments before performing the search. This `LeafWalker` instances are used throughout the search and keep track aggregation state as the search executes on tsid by tsid basis. But then `currentCompetitiveIterator` will not be updated, since  all leaf buckets have already been created.

I don't think this problem should block this change. The `composite` agg is already not working correctly with `TimeSeriesIndexSearcher`. I think for now we should add validation that prevents the usage of `time_series` aggregation with the `composite` aggregation. (and also for the `sampler` aggregation)

Reply:
> I don't think this problem should block this change. The composite agg is already not working correctly with TimeSeriesIndexSearcher. I think for now we should add validation that prevents the usage of time_series aggregation with the composite aggregation. (and also for the sampler aggregation)

Is that when the `time_series` agg is a parent of the composite or a child?

Reply:
This is when `time_series` agg is a sub aggregator of `composite` aggregator. I don't think that `composite` aggregator allows `time_series` aggregator to be its parent aggregator.

We will address this in elastic/elasticsearch#92390.

Reply:
Fair point. I'll remove it.